torch/_inductor/ir.py

```diff
@@ -2420,7 +2420,7 @@ def sorted_indices(arr):
         # reorder the stride given order
         stride_ordered = [-1] * len(order)
         for i in range(len(order)):
-            stride_ordered[order[i]] = V.graph.sizevars.size_hint(stride[i])
+            stride_ordered[order[i]] = V.graph.sizevars.size_hint(stride[i], fallback=2)
```

Comment:
This looks sus, it would have to be OK for `stride_ordered` to give incorrect results sometimes, which is super non-obvious from the method name and (non-existent) docs right now.

Reply:
@imzhuhl do you have the backtrace by which stride_ordered was called such that this was relevant handy?

Reply:
> @imzhuhl do you have the backtrace by which stride_ordered was called such that this was relevant handy?

Hi, I printed the backtrace using ipdb:
```bash
root@intel-spr32 ~/a/x/p/t/inductor (item_unsqueeze_support)# python test_torchinductor_dynamic_shapes.py -k test_item_unsqueeze_nobreak                                                                                          (torch_dev)
> /root/alibaba/xcode-ml/pytorch/torch/_inductor/ir.py(2423)is_stride_ordered()
   2422         import ipdb; ipdb.set_trace()
-> 2423         for i in range(len(order)):
   2424             stride_ordered[order[i]] = V.graph.sizevars.size_hint(stride[i], fallback=2)

ipdb> bt
  /root/alibaba/xcode-ml/pytorch/test/inductor/test_torchinductor_dynamic_shapes.py(539)<module>()
    537     # Slow on ASAN after https://github.com/pytorch/pytorch/pull/94068
    538     if (HAS_CPU or HAS_CUDA) and not TEST_WITH_ASAN:
--> 539         run_tests(needs="filelock")

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/test_case.py(41)run_tests()
     40                 return
---> 41     run_tests()
     42

  /root/alibaba/xcode-ml/pytorch/torch/testing/_internal/common_utils.py(1169)run_tests()
   1168     else:
-> 1169         unittest.main(argv=argv)
   1170

  /root/miniconda3/envs/torch_dev/lib/python3.10/unittest/main.py(101)__init__()
    100         self.parseArgs(argv)
--> 101         self.runTests()
    102

  /root/miniconda3/envs/torch_dev/lib/python3.10/unittest/main.py(271)runTests()
    270             testRunner = self.testRunner
--> 271         self.result = testRunner.run(self.test)
    272         if self.exit:

  /root/miniconda3/envs/torch_dev/lib/python3.10/unittest/runner.py(184)run()
    183             try:
--> 184                 test(result)
    185             finally:

  /root/miniconda3/envs/torch_dev/lib/python3.10/unittest/suite.py(84)__call__()
     83     def __call__(self, *args, **kwds):
---> 84         return self.run(*args, **kwds)
     85

  /root/miniconda3/envs/torch_dev/lib/python3.10/unittest/suite.py(122)run()
    121             if not debug:
--> 122                 test(result)
    123             else:

  /root/miniconda3/envs/torch_dev/lib/python3.10/unittest/suite.py(84)__call__()
     83     def __call__(self, *args, **kwds):
---> 84         return self.run(*args, **kwds)
     85

  /root/miniconda3/envs/torch_dev/lib/python3.10/unittest/suite.py(122)run()
    121             if not debug:
--> 122                 test(result)
    123             else:

  /root/miniconda3/envs/torch_dev/lib/python3.10/unittest/case.py(650)__call__()
    649     def __call__(self, *args, **kwds):
--> 650         return self.run(*args, **kwds)
    651

  /root/alibaba/xcode-ml/pytorch/torch/testing/_internal/common_device_type.py(477)run()
    476     def run(self, result=None):
--> 477         super().run(result=result)
    478         # Early terminate test if _stop_test_suite is set.

  /root/alibaba/xcode-ml/pytorch/torch/testing/_internal/common_utils.py(2833)run()
   2832             num_runs = MAX_NUM_RETRIES + 1 if RETRY_TEST_CASES else 1
-> 2833             self._run_with_retry(
   2834                 result=result,

  /root/alibaba/xcode-ml/pytorch/torch/testing/_internal/common_utils.py(2759)_run_with_retry()
   2758
-> 2759             super_run(result=result)
   2760

  /root/miniconda3/envs/torch_dev/lib/python3.10/unittest/case.py(591)run()
    590                     with outcome.testPartExecutor(self, isTest=True):
--> 591                         self._callTestMethod(testMethod)
    592                     outcome.expecting_failure = False

  /root/miniconda3/envs/torch_dev/lib/python3.10/unittest/case.py(549)_callTestMethod()
    548     def _callTestMethod(self, method):
--> 549         method()
    550

  /root/alibaba/xcode-ml/pytorch/torch/testing/_internal/common_utils.py(2667)wrapper()
   2666             with policy():
-> 2667                 method(*args, **kwargs)
   2668         return types.MethodType(wrapper, self)

  /root/alibaba/xcode-ml/pytorch/torch/testing/_internal/common_device_type.py(416)instantiated_test()
    415                     self._apply_precision_override_for_test(test, param_kwargs)
--> 416                     result = test(self, **param_kwargs)
    417                 except RuntimeError as rte:

  /root/miniconda3/envs/torch_dev/lib/python3.10/contextlib.py(79)inner()
     78             with self._recreate_cm():
---> 79                 return func(*args, **kwds)
     80         return inner

  /root/alibaba/xcode-ml/pytorch/test/inductor/test_torchinductor_dynamic_shapes.py(508)test_item_unsqueeze_nobreak()
    507
--> 508         f(torch.tensor([3, 4, 5, 6], device=device))
    509

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/eval_frame.py(501)_fn()
    500             try:
--> 501                 return fn(*args, **kwargs)
    502             finally:

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/eval_frame.py(667)catch_errors()
    666         with compile_lock, _disable_current_modes():
--> 667             return callback(frame, cache_entry, hooks, frame_state)
    668

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/convert_frame.py(383)_convert_frame_assert()
    382         with config.patch(_patch_config_if_changed()):
--> 383             compiled_product = _compile(
    384                 frame.f_code,

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/convert_frame.py(646)_compile()
    645         try:
--> 646             guarded_code = compile_inner(code, one_graph, hooks, transform)
    647             return guarded_code

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/utils.py(244)time_wrapper()
    243                 t0 = time.time()
--> 244                 r = func(*args, **kwargs)
    245                 time_spent = time.time() - t0

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/convert_frame.py(562)compile_inner()
    561             try:
--> 562                 out_code = transform_code_object(code, transform)
    563                 break

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/bytecode_transformation.py(1033)transform_code_object()
   1032
-> 1033     transformations(instructions, code_options)
   1034     return clean_and_assemble_instructions(instructions, keys, code_options)[1]

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/convert_frame.py(151)_fn()
    150         try:
--> 151             return fn(*args, **kwargs)
    152         finally:

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/convert_frame.py(527)transform()
    526             with tracing(tracer.output.tracing_context), tracer.set_current_tx():
--> 527                 tracer.run()
    528         except exc.UnspecializeRestartAnalysis:

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/symbolic_convert.py(2096)run()
   2095     def run(self):
-> 2096         super().run()
   2097

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/symbolic_convert.py(781)run()
    780                     and not self.output.should_exit
--> 781                     and self.step()
    782                 ):

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/symbolic_convert.py(744)step()
    743             )
--> 744             getattr(self, inst.opname)(inst)
    745

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/symbolic_convert.py(2211)RETURN_VALUE()
   2210         log.debug("RETURN_VALUE triggered compile")
-> 2211         self.output.compile_subgraph(
   2212             self,

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/output_graph.py(908)compile_subgraph()
    907             self.add_output_instructions(
--> 908                 self.compile_and_call_fx_graph(tx, list(reversed(stack_values)), root)
    909                 + [create_instruction("UNPACK_SEQUENCE", arg=len(stack_values))]

  /root/miniconda3/envs/torch_dev/lib/python3.10/contextlib.py(79)inner()
     78             with self._recreate_cm():
---> 79                 return func(*args, **kwds)
     80         return inner

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/output_graph.py(1076)compile_and_call_fx_graph()
   1075         with self.restore_global_state():
-> 1076             compiled_fn = self.call_user_compiler(gm)
   1077         compiled_fn = disable(compiled_fn)

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/utils.py(244)time_wrapper()
    243                 t0 = time.time()
--> 244                 r = func(*args, **kwargs)
    245                 time_spent = time.time() - t0

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/output_graph.py(1129)call_user_compiler()
   1128                 compiler_fn = WrapperBackend(compiler_fn)
-> 1129             compiled_fn = compiler_fn(gm, self.example_inputs())
   1130             _step_logger()(logging.INFO, f"done compiler function {name}")

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/repro/after_dynamo.py(117)debug_wrapper()
    116         else:
--> 117             compiled_gm = compiler_fn(gm, example_inputs)
    118

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/repro/after_dynamo.py(117)debug_wrapper()
    116         else:
--> 117             compiled_gm = compiler_fn(gm, example_inputs)
    118

  /root/alibaba/xcode-ml/pytorch/torch/__init__.py(1672)__call__()
   1671
-> 1672         return compile_fx(model_, inputs_, config_patches=self.config)
   1673

  /root/alibaba/xcode-ml/pytorch/torch/_inductor/compile_fx.py(1168)compile_fx()
   1167     ), compiled_autograd.disable():
-> 1168         return aot_autograd(
   1169             fw_compiler=fw_compiler,

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/backends/common.py(55)compiler_fn()
     54             with enable_aot_logging(), patch_config:
---> 55                 cg = aot_module_simplified(gm, example_inputs, **kwargs)
     56                 counters["aot_autograd"]["ok"] += 1

  /root/alibaba/xcode-ml/pytorch/torch/_functorch/aot_autograd.py(887)aot_module_simplified()
    886     with compiled_autograd.disable():
--> 887         compiled_fn = create_aot_dispatcher_function(
    888             functional_call,

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/utils.py(244)time_wrapper()
    243                 t0 = time.time()
--> 244                 r = func(*args, **kwargs)
    245                 time_spent = time.time() - t0

  /root/alibaba/xcode-ml/pytorch/torch/_functorch/aot_autograd.py(600)create_aot_dispatcher_function()
    599
--> 600         compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata)
    601         if aot_config.is_export:

  /root/alibaba/xcode-ml/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py(427)aot_wrapper_dedupe()
    426     if ok:
--> 427         return compiler_fn(flat_fn, leaf_flat_args, aot_config, fw_metadata=fw_metadata)
    428

  /root/alibaba/xcode-ml/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py(632)aot_wrapper_synthetic_base()
    631     if synthetic_base_info is None:
--> 632         return compiler_fn(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
    633

  /root/alibaba/xcode-ml/pytorch/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py(97)aot_dispatch_base()
     96             )
---> 97         compiled_fw = compiler(fw_module, updated_flat_args)
     98

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/utils.py(244)time_wrapper()
    243                 t0 = time.time()
--> 244                 r = func(*args, **kwargs)
    245                 time_spent = time.time() - t0

  /root/alibaba/xcode-ml/pytorch/torch/_inductor/compile_fx.py(1100)fw_compiler_base()
   1099
-> 1100         return inner_compile(
   1101             model,

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/repro/after_aot.py(83)debug_wrapper()
     82             # with fake inputs
---> 83             inner_compiled_fn = compiler_fn(gm, example_inputs)
     84         except Exception as e:

  /root/alibaba/xcode-ml/pytorch/torch/_inductor/debug.py(305)inner()
    304             with DebugContext():
--> 305                 return fn(*args, **kwargs)
    306

  /root/miniconda3/envs/torch_dev/lib/python3.10/contextlib.py(79)inner()
     78             with self._recreate_cm():
---> 79                 return func(*args, **kwds)
     80         return inner

  /root/alibaba/xcode-ml/pytorch/torch/_inductor/compile_fx.py(320)compile_fx_inner()
    319     else:
--> 320         compiled_graph = fx_codegen_and_compile(
    321             gm, example_inputs, **graph_kwargs  # type: ignore[arg-type]

  /root/alibaba/xcode-ml/pytorch/torch/_inductor/compile_fx.py(535)fx_codegen_and_compile()
    534         with V.set_graph_handler(graph):
--> 535             graph.run(*example_inputs)
    536             output_strides: List[Optional[Tuple[int, ...]]] = []

  /root/alibaba/xcode-ml/pytorch/torch/_dynamo/utils.py(244)time_wrapper()
    243                 t0 = time.time()
--> 244                 r = func(*args, **kwargs)
    245                 time_spent = time.time() - t0

  /root/alibaba/xcode-ml/pytorch/torch/_inductor/graph.py(518)run()
    517     def run(self, *args):
--> 518         return super().run(*args)
    519

  /root/alibaba/xcode-ml/pytorch/torch/fx/interpreter.py(138)run()
    137             try:
--> 138                 self.env[node] = self.run_node(node)
    139             except Exception as e:

  /root/alibaba/xcode-ml/pytorch/torch/_inductor/graph.py(868)run_node()
    867                         stride_order = ir.NHWC_STRIDE_ORDER
--> 868                     result = ir.ExternKernel.require_stride_order(result, stride_order)
    869

  /root/alibaba/xcode-ml/pytorch/torch/_inductor/ir.py(3714)require_stride_order()
   3713                 x.get_layout(), FixedLayout
-> 3714             ) and x.get_layout().is_stride_ordered(order):
   3715                 return x

> /root/alibaba/xcode-ml/pytorch/torch/_inductor/ir.py(2423)is_stride_ordered()
   2422         import ipdb; ipdb.set_trace()
-> 2423         for i in range(len(order)):
   2424             stride_ordered[order[i]] = V.graph.sizevars.size_hint(stride[i], fallback=2)

ipdb> p stride
[i3, 1]
ipdb> p order
[1, 0]
```

Reply:
OK. Potentially, you could modify graph.py to skip the stride order detection logic if it sees that the tensor has an unbacked symint size.

Reply:
Sorry, due to other commitments, I didn't get around to modifying this commit until today.

I modify graph.py to avoid entering stride order detection. Please help to check whether this modification is ok.